clang/lib/Sema/SemaChecking.cpp

```diff
@@ -179,6 +179,39 @@ static bool SemaBuiltinCHERICapCreate(Sema &S, CallExpr *TheCall) {
   return false;
 }
 
+/// Check that argument \p ArgIndex is a capability type (or an array/function
+/// that decays to a capability type.
+static bool checkCapArg(Sema &S, CallExpr *TheCall, unsigned ArgIndex,
+                        QualType *ResultingSrcTy = nullptr) {
+  clang::Expr *Source = TheCall->getArg(ArgIndex);
+  ASTContext &Ctx = S.Context;
+  QualType SrcTy = Source->getType();
+  // We should also be able to use it with arrays/functions
+  if (SrcTy->canDecayToPointerType())
+    SrcTy = Ctx.getDecayedType(SrcTy);
+
+  // Also allow NULL/nullptr in hybrid mode to be promoted to a capability
+  if (SrcTy->isNullPtrType())
+    SrcTy = Ctx.getPointerType(Ctx.VoidTy, ASTContext::PIK_Capability);
+  else if (SrcTy->isPointerType() && !SrcTy->isCapabilityPointerType() &&
+           Source->isNullPointerConstant(Ctx, Expr::NPC_ValueDependentIsNull))
+    SrcTy =
+        Ctx.getPointerType(SrcTy->getPointeeType(), ASTContext::PIK_Capability);
```

Comment:
I _think_ something like this is what you want (the getPointeeType makes it a bit tricky so I see now how it came to be the way it is):
```suggestion
  if (Source->isNullPointerConstant(Ctx, Expr::NPC_ValueDependentIsNull) &&
      !SrcTy->isCapabilityPointerType()) {
    if (SrcTy->isPointerType)
      SrcTy =
          Ctx.getPointerType(SrcTy->getPointeeType(), ASTContext::PIK_Capability);
    else /* Integer 0 or __null/nullptr */
      SrcTy = Ctx.getPointerType(Ctx.VoidTy, ASTContext::PIK_Capability);
  }
```

Reply:
Or, with a bit more DRY:
I _think_ something like this is what you want (the getPointeeType makes it a bit tricky so I see now how it came to be the way it is):
```suggestion
  if (Source->isNullPointerConstant(Ctx, Expr::NPC_ValueDependentIsNull) &&
      !SrcTy->isCapabilityPointerType()) {
    // Literal 0 and __null/nullptr get mapped to void * __capability
    auto PointeeTy = SrcTy->isPointerType() ? SrcTy->getPointeeType() : Ctx.VoidTy;
    SrcTy = Ctx.getPointerType(PointeeTy, ASTContext::PIK_Capability);
  }
```

Reply:
Seems like this is needed to handle `__null` anyway since that is defined to be an integer type for some reason.

The main reason I wanted to avoid 0 being converted to `(void* __capability)` is to handle cases where arguments were accidentally swapped.

```
__uintcap_t oops(__uintcap_t cap) {
  // clear purecap flag:
  return __builtin_cheri_flags_set(0, cap);
}
```
Fortunately this is caught by `warning: incompatible pointer to integer conversion returning 'void * __capability' from a function with result type '__uintcap_t' [-Wint-conversion]`, but I feel it would be nicer if zero wasn't converted from int -> void* __capability.

Additionally I was thinking of the builtin being more like an overloaded function which also gives an error for literal zero:
```
void* builtin_foo(void* cap, long arg) __attribute__((overloadable)); 
char* builtin_foo(char* cap, long arg) __attribute__((overloadable)); 
__intcap_t builtin_foo(__intcap_t cap, long arg) __attribute__((overloadable)); 
__uintcap_t builtin_foo(__uintcap_t cap, long arg) __attribute__((overloadable)); 

void test() {
    builtin_foo(0, 0);
}
```
This results in:
```
source>:7:5: error: call to 'builtin_foo' is ambiguous
    builtin_foo(0, 0);
    ^~~~~~~~~~~
<source>:1:7: note: candidate function
void* builtin_foo(void* cap, long arg) __attribute__((overloadable)); 
      ^
<source>:2:7: note: candidate function
char* builtin_foo(char* cap, long arg) __attribute__((overloadable)); 
      ^
<source>:3:12: note: candidate function
__intcap_t builtin_foo(__intcap_t cap, long arg) __attribute__((overloadable)); 
           ^
<source>:4:13: note: candidate function
__uintcap_t builtin_foo(__uintcap_t cap, long arg) __attribute__((overloadable)); 
            ^
1 error generated.
Compiler returned: 1
```

However, if you feel strongly that 0 should be allowed without casting it to a pointer type and changing the return value to void*, then I can allow it since I think all swapped argument cases *should* be caught by other diagnostics.